### PR TITLE
Dynamic require

### PR DESCRIPTION
**Pull request by danielgindi**
## Rollup Plugin Name: `commonjs`

This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [x] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

### Description

https://github.com/rollup/rollup-plugin-commonjs/issues/131
https://github.com/rollup/rollup-plugin-commonjs/pull/331

This has already been through major refactoring and adjustments, and now has been rebased on this repo.  
Json support has been fixed (and this will now required the json plugin when requiring a full module dynamically, as it needs to cache the `package.json`).  
Also the issue where the whole dynamic loader code was always added to any cjs project been rolled up is solved, it's now conditional on whether dynamic module loading is enabled or not.

Tests:  
There are 3 tests (`unambiguous-with-default-export`, `unambiguous-with-import`, `unambiguous-with-named-export`) which were slightly modified.  
The reasoning is that prior to this PR, if a file was detected to be using static `import`/`export`, then it was not processed for CJS `require` statements.  
That situation was a nice optimization but not realistic, as users coming from webpack are used to have a JS file with static `import`/`export`, while `import`ing or `require` a module which is essentially a CJS module. Such a module could have circular dependencies (i.e `logform`) or dynamic dependencies (i.e `knex`) which needs processing in the CJS realm.  
Now these 3 tests are using `require`, which means that there might be surprises hidden behind the scenes, so it must be processed for CJS. This means that the output code is a little more verbose (though not ending up adding the CJS boilerplate).

### CODE REVIEW

**Comment by shellscape** [comment]
Thanks for reopening this PR here. Please fix the CI errors and we'll take a look once that's working.

**Comment by danielgindi** [comment]
@shellscape Are the node-vxx-latest tests being done with latest version over PR tests? Some weird combination?

**Comment by shellscape** [comment]
Not sure I understand the question. Pretty standard setup: linting runs on the files in the PR and all tests for the plugin(s) changed run.

**Comment by danielgindi** [comment]
I'm asking because the tests on Node 8/10 windows - pass.  
Locally they pass on a mac with node 12.   
Analysis in the CI - passes.  
But node-vxx-latest tests do no pass, and they actually act like the new tests are running with the old code (which won't pass).

**Comment by danielgindi** [comment]
Found a potential cause for this. In the two tests that fail intermittently there was usage of `rollup-plugin-node-resolve`, which is not installed. I'm guessing it's installed somewhere on my local machine, and on one of the CI images, as sometimes it fails and sometimes does not.

**Comment by shellscape** [comment]
Well you should be leveraging @rollup/node-resolve, and you should make sure it's added properly to the package's devDeps. Monorepos work a little differently. If we don't have this outlined in the README or CONTRIBUTING let me know and we'll add it to the docs.

File: packages/commonjs/src/resolve-id.js [hunk 1]
```diff
@@ -3,11 +3,12 @@ import { statSync } from 'fs';
 import { dirname, resolve, sep } from 'path';
 
 import {
+  DYNAMIC_JSON_PREFIX, DYNAMIC_PACKAGES_ID,
```
Comment (shellscape) [inline packages/commonjs/src/resolve-id.js]: separate lines please

File: packages/commonjs/src/resolve-id.js [hunk 2]
```diff
@@ -3,11 +3,12 @@ import { statSync } from 'fs';
...
   getExternalProxyId,
   getIdFromProxyId,
   getProxyId,
   HELPERS_ID,
-  PROXY_SUFFIX
+  PROXY_SUFFIX,
```
Comment (shellscape) [inline packages/commonjs/src/resolve-id.js]: please remove trailing commas. not sure why linting didn't pick that up, but we don't have trailing commas enabled.

File: packages/commonjs/src/transform.js [hunk 3]
```diff
@@ -137,24 +145,65 @@ export function transformCommonjs(
   }
 
   function getRequired(node, name) {
-    const sourceId = getRequireStringArg(node);
+    let sourceId = getRequireStringArg(node);
+    const isDynamicRegister = sourceId.startsWith(DYNAMIC_REGISTER_PREFIX);
+    if (isDynamicRegister)
```
Comment (shellscape) [inline packages/commonjs/src/transform.js]: please wrap in brackets

File: packages/commonjs/src/transform.js [hunk 4]
```diff
@@ -137,24 +145,65 @@ export function transformCommonjs(
...
         } while (scope.contains(name));
       }
 
-      sources.push(sourceId);
-      required[sourceId] = { source: sourceId, name, importsDefault: false };
+      if (isDynamicRegister && sourceId.endsWith('.json')) {
+        sourceId = DYNAMIC_JSON_PREFIX + sourceId;
+      }
+
+      if (isDynamicRegister || !isDynamic || sourceId.endsWith('.json'))
```
Comment (shellscape) [inline packages/commonjs/src/transform.js]: please wrap in brackets

File: packages/commonjs/src/transform.js [hunk 5]
```diff
@@ -137,24 +145,65 @@ export function transformCommonjs(
...
     return required[sourceId];
   }
 
+  function hasDynamicModuleForPath(source) {
+    if (!/[/\\]/.test(source)) {
+      try {
+        const resolvedPath = normalizePathSlashes(
+          nodeResolveSync(source, { basedir: dirname(id) })
+        );
+        if (dynamicRequireModuleSet.has(resolvedPath)) return true;
```
Comment (shellscape) [inline packages/commonjs/src/transform.js]: please wrap in brackets

File: packages/commonjs/test/fixtures/form/ignore-ids-function/output.js [hunk 6]
```diff
@@ -1,7 +1,8 @@
+import * as commonjsHelpers from '_commonjsHelpers.js';
 import 'bar';
 import bar from '_bar?commonjs-proxy';
 
-var foo = require( 'foo' );
+var foo = commonjsHelpers.commonjsRequire( 'foo' ,"./fixtures/form/ignore-ids-function");
```
Comment (shellscape) [inline packages/commonjs/test/fixtures/form/ignore-ids-function/output.js]: We need a detailed explanation in the PR body as to why these fixtures were changed to such a degree. Typically when adding new features, the existing fixtures should be left as-is and tests should pass. If they don't, then that's a sign that the feature is breaking.  You've checked "no" for Breaking Changes, so these fixtures really shouldn't be modified.

File: packages/commonjs/test/types.ts [hunk 7]
```diff
@@ -14,7 +14,8 @@ const config: import("rollup").RollupOptions = {
 			ignoreGlobal: false,
 			sourceMap: false,
 			namedExports: { './module.js': ['foo', 'bar' ] },
-			ignore: [ 'conditional-runtime-dependency' ]
+			ignore: [ 'conditional-runtime-dependency' ],
+      dynamicRequireTargets: [ 'node_modules/logform/*.js' ]
```
Comment (shellscape) [inline packages/commonjs/test/types.ts]: spacing.

File: packages/commonjs/README.md [hunk 8]
```diff
@@ -127,6 +127,32 @@ Default: `[]`
 
 Sometimes you have to leave require statements unconverted. Pass an array containing the IDs or an `id => boolean` function. Only use this option if you know what you're doing!
 
+### `dynamicRequireTargets`
```
Comment (shellscape) [inline packages/commonjs/README.md]: options should be in alphabetic order

File: packages/commonjs/README.md [hunk 9]
```diff
@@ -127,6 +127,32 @@ Default: `[]`
...
+
+Type: `String|Array[String]`<br>
+Default: `[]`
+
+Some modules have dynamic `require` calls in them, or even circular dependencies (which are not handled well by EJS).
```
Comment (shellscape) [inline packages/commonjs/README.md]:
```suggestion
Some modules contain dynamic `require` calls, or require modules that contain circular dependencies, which are not handled well by [EJS](link to EJS info).  
```

What is EJS? Uncommon acronyms without links to info about them are generally unfriendly to new users. Please add a link to info about EJS.

File: packages/commonjs/README.md [hunk 10]
```diff
@@ -127,6 +127,32 @@ Default: `[]`
 
 Sometimes you have to leave require statements unconverted. Pass an array containing the IDs or an `id => boolean` function. Only use this option if you know what you're doing!
 
+### `dynamicRequireTargets`
+
+Type: `String|Array[String]`<br>
+Default: `[]`
+
+Some modules have dynamic `require` calls in them, or even circular dependencies (which are not handled well by EJS).  
+Including those modules as `dynamicRequireTargets` will simulate a CJS environment for them with support for dynamic and circular dependencies.
```
Comment (shellscape) [inline packages/commonjs/README.md]: Please don't use hard link breaks for paragraphs unless two or more lines are needed between them. Backspace so this line wraps naturally. We're big on getting our docs pristine here (even though we have a ways to go yet)

File: packages/commonjs/README.md [hunk 11]
```diff
@@ -127,6 +127,32 @@ Default: `[]`
...
 Sometimes you have to leave require statements unconverted. Pass an array containing the IDs or an `id => boolean` function. Only use this option if you know what you're doing!
 
+### `dynamicRequireTargets`
+
+Type: `String|Array[String]`<br>
+Default: `[]`
+
+Some modules have dynamic `require` calls in them, or even circular dependencies (which are not handled well by EJS).  
+Including those modules as `dynamicRequireTargets` will simulate a CJS environment for them with support for dynamic and circular dependencies.  
+Note that this might hardcode some paths from your PC in the final js file, so you might want to search & replace the base path of the root js file.
```
Comment (shellscape) [inline packages/commonjs/README.md]:
```suggestion

_Note: This feature may result in some paths being rendered as absolute in the final bundle. That may require replacing ...<please explain more>._
```

I'm not quite sure what this note is supposed to tell the user they might have to do. I've suggested a change to the note for clearer initial language, but we nee to explain better to users what's up.

File: packages/commonjs/test/types.ts (hunk 7 above)
Comment (danielgindi) [reply packages/commonjs/test/types.ts]: It's weird some code in the project has tabs, some spaces. Here it's tabs.

File: packages/commonjs/README.md (hunk 9 above)
Comment (danielgindi) [reply packages/commonjs/README.md]: Wrong teminology.

**Comment by danielgindi** [comment]
@shellscape Did some cleaning. I think it's ready as it already went through a lot of evolution (with the help of @lukastaegert), and it's a major part of what people are expecting when rolling up common js modules.
Also it's getting hard to rebase every time that the whole project changes (prettier, ava, repo change...).

File: packages/commonjs/test/fixtures/form/ignore-ids-function/output.js (hunk 6 above)
Comment (danielgindi) [reply packages/commonjs/test/fixtures/form/ignore-ids-function/output.js]: I remember there was a very valid reason, but I'll look into it again to make sure I know what I'm talking about when I'm talking about it :-)

**Comment by shellscape** [comment]
> Also it's getting hard to rebase every time that the whole project changes

There's only been one major change so far, and that was for the migration to the monorepo. You shouldn't see any tooling changes for a long while. 

We'll need resolution to https://github.com/rollup/plugins/pull/206#discussion_r378235133 before we can move forward as well. I'm a stickler for making sure big changes are properly documented. 

_An aside: We've both ESLint and Prettier setup in the repo. I'd very highly recommend a plugin for your editor of choice which will automatically correct formatting/style on document save. It takes a lot of headache out of complying with linting guidelines._

File: packages/commonjs/test/fixtures/form/ignore-ids-function/output.js (hunk 6 above)
Comment (danielgindi) [reply packages/commonjs/test/fixtures/form/ignore-ids-function/output.js]: Okay, reverted it for 3 tests, and another 3 tests are problematic for a valid reason. Adding notes to the PR.

**Comment by shellscape** [comment]
Thanks. I'll bug @lukastaegert to take another look.

**Comment by lukastaegert** [comment]
Ok, this one is huge and I cannot promise when I will find time to look into this as my current stack of things to do is really long. One thing, I did not see any form tests that would allow me to inspect what the actually generated code is in the PR. Except deleted ones. Would it be possible to add some? I would not be too happy if the PR produced noticeably less efficient code for situations where we do not have dynamic requires.

**Comment by danielgindi** [comment]
@lukastaegert So are you suggesting restoring the 3 deleted ones and updating the expected outputs?

**Comment by lukastaegert** [comment]
That would definitely help with the review and might also prevent regressions in code quality. Why were they deleted in the first place?

**Comment by lukastaegert** [comment]
Alternatively, I think it would be even more interesting to see what Rollup makes of it in the end. There are already some function tests that do that via `getCodeFromBundle`, e.g. "produces optimized code when importing esm without a default export". But maybe we can do this as well for sample based tests. Let me think.

**Comment by lukastaegert** [comment]
Ok, here is a really simple change that will require little work of you but make the test much more powerful (and might even help us get rid of the explicit form tests eventually): Change `test/function.js` and add `t.snapshot(code);` as the last assertion for each test (i.e. after `if (config.global) config.global(global, t);`). Then each function test will also generate a snapshot that we can compare. I think this will also help with future reviews to see what changed.

**Comment by danielgindi** [comment]
> That would definitely help with the review and might also prevent regressions in code quality. Why were they deleted in the first place?

Outlined in the PR description

**Comment by danielgindi** [comment]
I am not that (or at all) familiar with `ava` so I prefer not do change the tests in a way that I do not fully understand.

Anyway I found the root of the issue, brought back the tests with only a minor modification, and I'm updating the PR description.

**Comment by danielgindi** [comment]
@shellscape Any news on this?

File: packages/commonjs/test/snapshots/function.js.md [hunk 12]
```diff
@@ -215,6 +215,2082 @@ Generated by [AVA](https://ava.li).
...
+    	__proto__: null,␊
+    	commonjsGlobal: commonjsGlobal,␊
+    	unwrapExports: unwrapExports,␊
+    	createCommonjsModule: createCommonjsModule,␊
+    	getCjsExportFromNamespace: getCjsExportFromNamespace,␊
+    	commonjsRegister: commonjsRegister,␊
+    	commonjsRequire: commonjsRequire␊
+    });␊
+    ␊
+    _commonjsHelpers.commonjsRegister("/Users/lukastaegert/test/dynamic-require-plugins/packages/commonjs/test/fixtures/function/dynamic-require/submodule.js", function (module, exports) {␊
```
Comment (lukastaegert) [inline packages/commonjs/test/snapshots/function.js.md]: This is not acceptable to me. IMO is no reason to use absolute paths. Instead, you could determine the common root directory of all dynamic require targets and make all paths relative to that. E.g. use https://www.npmjs.com/package/commondir

File: packages/commonjs/src/index.js [hunk 13]
```diff
@@ -156,24 +250,56 @@ export default function commonjs(options = {}) {
         });
       }
 
+      // TODO: For code splitting, the runtime probably needs to be imported by each entry point
```
Comment (lukastaegert) [inline packages/commonjs/src/index.js]: Probably? This is one of the two most important plugins of the ecosystem. Why not write a test and if it does not work, address this? This is important!

Comment (danielgindi) [reply packages/commonjs/src/index.js]: @lukastaegert This comment is by you, after you did some refactoring on the original repo.

Could you help with a test for this one? I've never tried codesplitting in rollup, no experience with it. Don't want any mistakes there.

File: packages/commonjs/test/snapshots/function.js.md (hunk 12 above)
Comment (danielgindi) [reply packages/commonjs/test/snapshots/function.js.md]: I thought about adding such a function, but did not want to make the PR larger than it needs to be, or add more dependencies. If you're OK with this, then I'll roll with you.

Anyway the comment in the docs should stay, as the "common dir" could still expose private directory names.

**Comment by danielgindi** [comment]
> Just two very basic comments that need to be addressed if I should give my "ok" for this. Also, I did not do a full code review and I am not sure I will find the time to do it as it is a massive amount of changes and also I find the code very hard to understand.

I understand this is massive and takes time. However I think this is a fundamental feature, especially when a lot of people are migrating from Webpack (where everything always works under a virtualized commonjs ecosystem).


> This plugin always had code-quality issues but it does not get better if you blow up functions to many hundreds of lines of code. After all, someone needs to maintain this once it is released, and the first thing to maintaining it is to understand what is going on. Extracting some well-named helper functions and possibly moving them to different files grouped by what they are concerned with would go a long way to improve this.

Well, we already did all of that. I don't know if you recalled, but you went through pieces of code that you did not fully understand, and we extracted, renamed, moved things around.  
You cannot neutralize the complexity completely, without removing the feature completely.  
We can only refactor to make things more readable to other maintainers, but getting feedback from maintainers.

> Sorry I have to be so blunt.

Imagine I've already put up with all the refactoring and rebasing of this PR in the past :-)  
I do this now not out of necessity as I just ref the specific branch, but for the community. To give back. It would be a shame for people to not have this feature as this locks them in the webpack realm.

> I know you already put a lot of effort in this and this has been in the works for a long time. But code-quality was also the main reason why this did not really advance the first time and this really needs to improve. Also, I have less time on my hands these days.

Understandably. I appreciate your co-work on this, and I have even less time than I ever had, doing this at midnight or replying from the phone while on the... nevermind.

**Comment by shellscape** [comment]
@lukastaegert with v2.0.0 out the door, would you have any time to circle back to this?

@danielgindi is there any chance you know of someone you would be willing to nominate for caretaker of this plugin? I think lukas' comments about having to maintain it were pertinent because there's currently no one on the project on a regular basis that is intimately familiar with the plugin, and it's one of two of the most complex plugins. I get where both of you are coming from.

**Comment by danielgindi** [comment]
@shellscape I think that after resolving the last comment (about code splitting), which I do not even know how to test, we should be ready with tests that make sure nothing breaks along the way.
After that - I could maintain it when necessity arises. I did not catch a COVID-19 or something, so I’m still around 😬

**Comment by lukastaegert** [comment]
Hi, sorry my earlier comments were a little out of place. I will get back to this on Monday to add some code-splitting test and have another look @shellscape thanks for moderating!

**Comment by lukastaegert** [comment]
I refactored the test runner so that it supports code-splitting. While doing that, I discovered there already is a bug in this plugin that you cannot require an entry point that is itself commonjs, but this is unrelated. I added a test case that does not need this, but I could not get it to run because it did not find my import targets, maybe you can have a look.

**Comment by danielgindi** [comment]
Found the problem, right there in the TODO comment. This originated in my lack of familiarity with the code-splitting in rollup, even more back then.

**Comment by danielgindi** [comment]
@lukastaegert Do you know what the issue is with the ci throwing `Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up-to-date with packages/commonjs/package.json`?

**Comment by shellscape** [comment]
Amazing work here fellas!

**Comment by shaodahong** [comment]
Can resolve webpack `require.context`?

**Comment by danielgindi** [comment]
@shaodahong It’s definitely not implemented here, and I’m not sure about it anyway:

1. This is a webpack specific feature, not being compatible in anyway with a “clean” node environment.

2. Rollup in general works in a completely different way. It tries to convert everything to *static* imports/exports, and when the code is static- it’s much simpler doing optimizations. So this feature here in this plugin takes care only of specific cases when a dynamic `require` is still necessary (which is relatively rare), which means most if the modules won't show up in `require.context` anyway.

3. Webpack generates a dynamic module mapping for everything, wrapping every module import/export in the virtual fs it has for `require` statements. This is the *exact opposite* of what rollup does.

4. Webpack has to generate a comolete map in memory if evverything before generating the code. This allows it to supply the compile-time `require.context`, while rollup does not work like this. It has a plugin system, this plugin is one of them, and each plugin takes care of pieces of code on demand, and does not control the whole code. So it can’t generate the same module map in the same way.

My conclusion here is that:

1. The use-case for `require.context` needs to be investigated (with more people possibly using it weighing in here).
2. If it will seem necessary, it should probably implemented in the main rollup module, as a different feature, that has nothing to do with `require` but with all of the modules.

**Comment by shaodahong** [comment]
@danielgindi Thanks for your reply

**Comment by shellscape** [comment]
@danielgindi Apologies that I did not get to merging this before a conflict happened. Please run the following commands in the project to resolve that conflict:

```console
git checkout master
git reset --hard origin && git pull
git checkout dynamic_require
git merge master
rm pnpm-lock.yaml
pnpm install --force
git add pnpm-lock.yaml
git commit --no-edit
git push
```

And you should be good. Every now and then the pnpm lockfile gets out of sorts and that's the fix for a conflict.

**Comment by danielgindi** [comment]
> @danielgindi Apologies that I did not get to merging this before a conflict happened. Please run the following commands in the project to resolve that conflict:

Done!

**Comment by danielgindi** [comment]
@shellscape Don’t you want to merge it before breaking it again? 😂

**Comment by shellscape** [comment]
Gah. Done. Our day care closed so things are a bit bananas around here now.

**Comment by danielgindi** [comment]
> Gah. Done. Our day care closed so things are a bit bananas around here now. 

Ours too, two weeks ago. It’s double bananas here. The girls are climbing on me all day, can’t do any serious work.

**Comment by justin808** [comment]
Any link to how to use the master branch to use this plugin? 

Or any estimate on the publishing of the next version?

**Comment by mesqueeb** [comment]
I just got this error when trying to use a library of mine in which I use dynamic require:
```
Error: Dynamic requires are not currently supported by @rollup/plugin-commonjs
```

Should this have already been resolved by this thread? I am on
```
"@rollup/plugin-node-resolve": "^6.1.0",
```

**Comment by danielgindi** [comment]
@mesqueeb Please read the [README](https://github.com/rollup/plugins/tree/master/packages/commonjs). You need to use the latest version, and configure the `dynamicRequireTargets` option.

**Comment by mesqueeb** [comment]
@danielgindi Thank you very much!!!

**Comment by mjmnagy** [comment]
> `dynamicRequireTargets`

Ive been trying to get this going per your comment and have had no success?

if i have a require statement like to:
`this.svg = require('@icn/${set}/${icon}.svg')`

would i need to list each possibility for set and icon?
ex:
@icn/brands/facebook.svg
@icn/light/lemon.svg
@icn/thick/lemon.svg

Thanks

**Comment by danielgindi** [comment]
@mjmnagy `require('@icn/${set}/${icon}.svg')` is a static require, not a dynamic one. But ``require(`@icn/${set}/${icon}.svg`)`` is dynamic.
Anyway yes you should pass a glob pattern that matches all modules that you want to be available for the dynamic require.

**Comment by mjmnagy** [comment]
> @mjmnagy `require('@icn/${set}/${icon}.svg')` is a static require, not a dynamic one. But `` require(`@icn/${set}/${icon}.svg`) `` is dynamic.
> Anyway yes you should pass a glob pattern that matches all modules that you want to be available for the dynamic require.

Sorry for the confusion - it was a translation error. The `require` is dynamic and I have provided the `dynamicRequireTargets` with a listing of files paths. Where should they be referenced from `node_modules`? 

The function I have created to loop through the icons folders and create an array of strings of paths is::
```

const fs = require('fs')
const path = require('path')

const getAllFiles = function(dirPath, arrayOfFiles) {
  let dir = path.join(__dirname, 'dist', '/')
  let files = fs.readdirSync(dirPath)
  arrayOfFiles = arrayOfFiles || []

  files.forEach(function(file) {
    if (fs.statSync(dirPath + '/' + file).isDirectory()) {
      arrayOfFiles = getAllFiles(dirPath + '/' + file, arrayOfFiles)
    } else {
      arrayOfFiles.push(
        path
          .join(/*'node_modules', 'icons', */ '/', 'dist', dirPath, '/', file) //? not sure where the path should be from
          .replace(dir, '')
      )
    }
  })

  return arrayOfFiles
}

export default getAllFiles
```
No luck with this approach as I get `Dynamic requires are not supported...`
Any ideas would be greatly appreciated

**Comment by danielgindi** [comment]
```
dynamicRequireTargets: [
  'node_modules/**/*.svg'
]
```

**Comment by danielgindi** [comment]
Please just read the docs, also look at issues or open an issue. This PR is tested, merged, and working in production.

**Comment by danielgindi** [comment]
@lukastaegert please lock the conversation...

**Comment by mjmnagy** [comment]
> Please just read the docs, also look at issues or open an issue. This PR is tested, merged, and working in production.

I have and I could not get 

dynamicRequireTargets: [
  'node_modules/**/*.svg'
]

to work hence why I'm asking

**Pull request merged**